contracts/contracts/MessageProcessor.sol

```diff
@@ -82,51 +84,36 @@ contract MessageProcessor is Ownable, SnarkCommon, Hasher, CommonUtilities, IMes
       revert NoMoreMessages();
     }
 
-    // The state AccQueue must be merged
-    if (!poll.stateMerged()) {
-      revert StateNotMerged();
-    }
-
     // Retrieve stored vals
-    (, uint8 messageTreeSubDepth, uint8 messageTreeDepth, uint8 voteOptionTreeDepth) = poll.treeDepths();
+    (, uint8 voteOptionTreeDepth) = poll.treeDepths();
     // calculate the message batch size from the message tree subdepth
-    uint256 messageBatchSize = TREE_ARITY ** messageTreeSubDepth;
-
-    (, AccQueue messageAq) = poll.extContracts();
-
-    // Require that the message queue has been merged
-    uint256 messageRoot = messageAq.getMainRoot(messageTreeDepth);
-    if (messageRoot == 0) {
-      revert MessageAqNotMerged();
-    }
+    uint256 messageBatchSize = poll.batchSizes();
 
     // Copy the state and ballot commitment and set the batch index if this
     // is the first batch to process
     if (numBatchesProcessed == 0) {
       uint256 currentSbCommitment = poll.currentSbCommitment();
       sbCommitment = currentSbCommitment;
-      (, uint256 numMessages) = poll.numSignUpsAndMessages();
-      uint256 r = numMessages % messageBatchSize;
 
-      currentMessageBatchIndex = numMessages;
+      poll.padLastBatch();
+      batchHashes = poll.getBatchHashes();
+      currentBatchIndex = batchHashes.length;
 
-      if (currentMessageBatchIndex > 0) {
-        if (r == 0) {
-          currentMessageBatchIndex -= messageBatchSize;
-        } else {
-          currentMessageBatchIndex -= r;
-        }
+      if (currentBatchIndex > 0) {
+        currentBatchIndex -= 1;
       }
     }
 
+    uint256 inputBatchHash = batchHashes[currentBatchIndex - 1];
+    uint256 outputBatchHash = batchHashes[currentBatchIndex];
+
     if (
       !verifyProcessProof(
-        currentMessageBatchIndex,
-        messageRoot,
+        currentBatchIndex,
+        outputBatchHash,
         sbCommitment,
         _newSbCommitment,
-        messageTreeSubDepth,
-        messageTreeDepth,
+        uint8(messageBatchSize),
```

Comment:
this is dangerous as it could overflow during casting if the value is > 255 - maybe better to set this as uint8 from the start (poll contract) as probably never gonna have > 255 batch given circuit constraints

Reply:
resolved